File: Data Structures & Algorithms/sliding-window-maximum/submission-5.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        q = deque() # [0, 1, ]
        res = []
        for i in range(len(nums)):# i = 0, nums[0] = 1
            while q and i - k >= q[0]:
                q.popleft()
            while q and nums[q[-1]] < nums[i]:
                q.pop()
            q.append(i)
            if i >= k - 1:
                res.append(nums[q[0]])
        return res

        

File: Data Structures & Algorithms/top-k-elements-in-list/submission-5.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        
        int n = nums.length;
        Deque<Integer>[] buckets = new LinkedList[n + 1];

        Map<Integer, Integer> counts = new HashMap<>();

        for (int num : nums)
            counts.put(num, counts.getOrDefault(num, 0) + 1);

        counts.forEach((num, freq) -> {
            if (buckets[freq] == null)
                buckets[freq] = new LinkedList<>();
            buckets[freq].offer(num);
        });

        int i = n;
        int[] res = new int[k];
        int index = 0;
        while (i >= 0) {
            while ((buckets[i] != null) && !buckets[i].isEmpty()) {
                res[index++] = buckets[i].poll();
                k--;
                if (k == 0)
                    return res;
            }
            i--;
        }

        return new int[] {};

    }
}
